pkg/controller/constrainttemplate/constrainttemplate_controller_test.go

```diff
@@ -211,15 +222,48 @@ violation[{"msg": "denied!"}] {
 		Name:      "FooNamespace",
 		Object:    runtime.RawExtension{Object: ns},
 	}
-	resp, err := opa.Review(context.TODO(), req)
+	resp, err := opa.Review(ctx, req)
 	g.Expect(err).NotTo(gomega.HaveOccurred())
 	if len(resp.Results()) != 1 {
 		fmt.Println(resp.TraceDump())
-		fmt.Println(opa.Dump(context.TODO()))
+		fmt.Println(opa.Dump(ctx))
 	}
 	g.Expect(len(resp.Results())).Should(gomega.Equal(1))
 
+	// If the CRD is deleted out from underneath the template, it is recreated
+	log.Info("TESTING CRD RECREATION ON DELETE")
+	crd := &apiextensionsv1beta1.CustomResourceDefinition{}
+	crd.SetName(crdKey.Name)
+	crd.Spec = apiextensionsv1beta1.CustomResourceDefinitionSpec{}
+	g.Expect(c.Delete(ctx, crd)).NotTo(gomega.HaveOccurred())
+
+	g.Eventually(func() error {
+		// Drain the request queue so the controller can operate
+		select {
+		case v := <-requests:
+			log.Info("Reconciling request", "request", v)
+		default:
+		}
+		crd := &apiextensionsv1beta1.CustomResourceDefinition{}
+		if err := c.Get(ctx, crdKey, crd); err != nil {
+			return err
+		}
+		if !crd.GetDeletionTimestamp().IsZero() {
```

Comment:
There's a race if c is a caching client, correct? Perhaps we should comment above on why we're creating a new client (working around the RESTMapper caching). Perhaps we should compare UIDs of old vs. new to ensure we're dealing with a recreated CRD, in case the caching assumption changes?

Reply:
Hmm - looks like I'm behind the times. `DynamicRESTMapper` was [added](https://github.com/kubernetes-sigs/controller-runtime/pull/554) to controller-runtime back in 0.3.0. It was set to the default [recently](https://github.com/kubernetes-sigs/controller-runtime/commit/7dfd3c9af3f4357df4b8eac922f2767198d81d3d), but is as of yet unreleased. That said, it's simple to configure the manager to use it.

Reply:
Configuring DynamicRestMapper would be a huge boon. There may still be an issue with resources disappearing out from under us as CRDs are deleted, I'm not sure how that would work, but it'd be nice not to have to refresh the client so often.

What is the race condition in this instance? My understanding is that the client cache is caching the type-specific client, which maps to a specific URI endpoint. IIUC the lifecycle should go: resource deleting, error: missing URI not found, CRD is established and the resource is being served again.

Reply:
Manager has two different clients - the one returned by `mgr.GetClient()` will by default delegate to an informer cache, so the race is writing and then expecting to see that immediately reflected in a read which has not yet reached the informer.

In contrast, `mgr.GetAPIReader()` is a "direct path" to the API Server.

Reply:
Gah, ya. I missed that other client creation path. comparing UUIDs is a good idea. It looks like we are already defaulting to the new dynamic rest mapper, which is good news.

Reply:
Switched to using UUIDs